core/src/main/java/org/infinispan/affinity/impl/KeyAffinityServiceImpl.java

```diff
@@ -104,20 +107,15 @@ public K getKeyForAddress(Address address) {
          throw new NullPointerException("Null address not supported!");
 
       BlockingQueue<K> queue = null;
-      maxNumberInvariant.readLock().lock();
-      try {
-         queue = address2key.get(address);
-         if (queue == null)
-            throw new IllegalStateException("Address " + address + " is no longer in the cluster");
-      } finally {
-         maxNumberInvariant.readLock().unlock();
-      }
       try {
          K result = null;
          while (result == null && !keyGenWorker.isStopped()) {
             // obtain the read lock inside the loop, otherwise a topology change will never be able
             // to obtain the write lock
             maxNumberInvariant.readLock().lock();
+            queue = address2key.get(address);
+            if (queue == null)
+               throw new IllegalStateException("Address " + address + " is no longer in the cluster");
```

Comment:
These lines should be inside the try block, otherwise the lock may not be released.